optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/locator/EqualsLocationStrategy.java

```diff
@@ -22,7 +22,11 @@
 
     @Override
     public void addWorkingObject(Map<Object, Object> idToWorkingObjectMap, Object workingObject) {
-        idToWorkingObjectMap.put(workingObject, workingObject);
+        Object oldAddedObject = idToWorkingObjectMap.put(workingObject, workingObject);
+        if (oldAddedObject != null) {
+            throw new IllegalStateException("An object that is equal to workingObject (" + workingObject
+                    + ") has already been added.");
+        }
```

Comment:
+1, but the error message should include the toString() of both the old and the new one. Adjusting that now.